File: elixir/3.exs
# What is the largest prime factor of the number 600851475143?

defmodule Euler do

  def factor(n) when n < 2, do: []

  def factor(n) do
    d = 2..trunc(:math.ceil(:math.sqrt(n))) # speeds things up
      |> Stream.filter(&(rem(n, &1) == 0))
      |> Enum.to_list
      |> List.first || n # if no factor found, return itself

    [d] ++ factor(trunc(n/d))
  end
end

IO.puts List.last(Euler.factor(600851475143)) # 6857

File: ruby/3.rb
# What is the largest prime factor of the number 600851475143? <<<--->>>

def factorise(n)
  case n
  when 3
    [1, 3]
  when 5
    [1, 5]
  when 7
    [1, 7]
  else
    d = (2..Math.sqrt(n).ceil).find { |e| n % e == 0 }
    d = n if d.nil?
    [n/d, d]
  end
end

def lpf(n)
  composites = [n]
  factors = []
  until composites.empty?
    new_n, d = factorise(composites.pop)
    factors << d
    composites << new_n unless new_n == 1
  end
  factors.last
end

puts lpf(600851475143) # 6857